exonum/src/runtime/mod.rs

```diff
@@ -198,37 +198,42 @@ pub trait Runtime: Send + Debug + 'static {
     /// * Ensure that the deployed artifact has the following information, even if it is empty.
     fn artifact_protobuf_spec(&self, id: &ArtifactId) -> Option<ArtifactProtobufSpec>;
 
-    /// Restart previously added service instance with the given specification.
+    /// Runs the constructor of a new service instance with the given specification
+    /// and initial configuration.
+    ///
+    /// The service is not guaranteed to be added to the runtime at this point, so the runtime
+    /// may freely discard the instantiated service instance after completing this method. To
+    /// add the service to the runtime permanently, `add_service()` is called.
     ///
     /// # Policy on Panics
     ///
     /// * Catch each kind of panics except for `FatalError` and convert
     /// them into `ExecutionError`.
-    /// * If panic occurs, the runtime must ensure that it is in a consistent state.
-    fn restart_service(&mut self, spec: &InstanceSpec) -> Result<(), ExecutionError>;
+    fn start_adding_service(
+        &self,
+        context: ExecutionContext,
+        spec: &InstanceSpec,
+        parameters: Vec<u8>,
+    ) -> Result<(), ExecutionError>;
 
-    /// Add a new service instance with the given specification and initial configuration.
+    /// Permanently adds a service to the runtime.
     ///
-    /// The configuration parameters passed to the method are discarded immediately.
-    /// So the service instance should save them by itself if it is important for
-    /// the service business logic.
+    /// It is guaranteed that `start_adding_service()` was called with the same `spec` earlier
+    /// and returned `Ok(())`. Note that this previous call may have happened indefinite time ago;
+    /// indeed, the method is called for all services on the node startup.
     ///
     /// # Policy on Panics
```

Comment:
Not really a policy on panics, but rather on errors. Should runtime implementation catch all panics or will Core do it itself?